File: src/path.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
MIT License

Copyright (c) 2024 cauchy1988

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""


from __future__ import annotations
from typing import List, Tuple, Optional
from meta import Meta


class PathError(Exception):
	"""Custom exception for Path operations"""
	pass


class Path:
	def __init__(self, vehicle_id: int, meta_obj: Meta, need_init=True):
		self.meta_obj = meta_obj
		self.vehicle_id = vehicle_id
		
		if need_init:
			# Check if vehicle and its nodes still exist
			if vehicle_id not in self.meta_obj.vehicles:
				raise PathError(f"Vehicle {vehicle_id} not found in meta_obj")
			
			# node id route
			start_node_id = self.meta_obj.vehicles[self.vehicle_id].start_node_id
			end_node_id = self.meta_obj.vehicles[self.vehicle_id].end_node_id
			
			if start_node_id not in self.meta_obj.nodes or end_node_id not in self.meta_obj.nodes:
				raise PathError(f"Vehicle {vehicle_id} nodes not found: start={start_node_id}, end={end_node_id}")
			
			self.route: List[int] = [start_node_id, end_node_id]
			
			# start the service time along the node route
			earliest_time = self.meta_obj.nodes[start_node_id].earliest_service_time
			arrival_time = earliest_time + self.meta_obj.nodes[start_node_id].service_time + self.meta_obj.vehicle_run_between_nodes_time[self.vehicle_id][start_node_id][end_node_id]
			latest_time = max(arrival_time, self.meta_obj.nodes[end_node_id].earliest_service_time)
			
			if latest_time > self.meta_obj.nodes[end_node_id].latest_service_time:
				raise PathError(f"Time window violation: {latest_time} > {self.meta_obj.nodes[end_node_id].latest_service_time}")
			
			self.start_service_time_line: List[float] = [earliest_time, latest_time]
			
			# capacity load line
			# load after each node in route
			self.load_line: List[float] = [self.meta_obj.nodes[start_node_id].load,
			                  self.meta_obj.nodes[start_node_id].load + self.meta_obj.nodes[end_node_id].load]
			
			# distance accumulated
			end_distance = self.meta_obj.distances[start_node_id][end_node_id]
			self.distances: List[float] = [0, end_distance]
			
	def copy(self):
		new_path = Path(self.vehicle_id, self.meta_obj, False)
		new_path.meta_obj = self.meta_obj
		new_path.vehicle_id = self.vehicle_id
		new_path.route = self.route.copy()
		new_path.start_service_time_line = self.start_service_time_line.copy()
		new_path.load_line = self.load_line.copy()
		new_path.distances = self.distances.copy()
		
		return new_path
	
	def is_path_free(self):
		return len(self.route) <= 2
	
	@property
	def whole_time_cost(self):
		return self.start_service_time_line[-1] - self.start_service_time_line[0]

	@property
	def whole_distance_cost(self):
		return self.distances[-1] if self.distances else 0.0

	def _update_service_times_after_insertion(self, start_idx: int) -> bool:
		"""Update service times after inserting nodes at start_idx"""
		for i in range(start_idx, len(self.start_service_time_line)):
			prev_node_id = self.route[i - 1]
			current_node_id = self.route[i]
			new_start_time = max(self.start_service_time_line[i - 1] +
							    self.meta_obj.nodes[prev_node_id].service_time +
			                    self.meta_obj.vehicle_run_between_nodes_time[self.vehicle_id][prev_node_id][current_node_id],
			                    self.meta_obj.nodes[current_node_id].earliest_service_time)
			
			if new_start_time > self.meta_obj.nodes[current_node_id].latest_service_time:
				return False
			self.start_service_time_line[i] = new_start_time
		return True

	def _update_loads_after_insertion(self, pick_idx: int, delivery_idx: int) -> bool:
		"""Update loads after inserting pickup and delivery nodes"""
		for i in range(pick_idx, delivery_idx + 1):
			current_node_id = self.route[i]
			new_load = self.load_line[i - 1] + self.meta_obj.nodes[current_node_id].load
			if new_load > self.meta_obj.vehicles[self.vehicle_id].capacity:
				return False
			self.load_line[i] = new_load
		return True

	def _update_distances_after_insertion(self, start_idx: int):
		"""Update distances after inserting nodes at start_idx"""
		for i in range(start_idx, len(self.distances)):
			prev_node_id = self.route[i - 1]
			current_node_id = self.route[i]
			current_distance = self.distances[i - 1] + \
							   self.meta_obj.distances[prev_node_id][current_node_id]
			self.distances[i] = current_distance

	def try_to_insert_request(self, request_id: int, pick_insert_idx: int, delivery_insert_idx: int) -> Tuple[bool, float, float]:
		"""Try to insert a request into the path"""
		if pick_insert_idx >= delivery_insert_idx:
			return False, 0.0, 0.0
			
		pick_node_id = self.meta_obj.requests[request_id].pick_node_id
		delivery_node_id = self.meta_obj.requests[request_id].delivery_node_id
		
		# Insert nodes
		self.route.insert(pick_insert_idx, pick_node_id)
		self.route.insert(delivery_insert_idx, delivery_node_id)
		
		# Insert placeholder values
		prev_whole_time_cost = self.whole_time_cost
		self.start_service_time_line.insert(pick_insert_idx, 0)
		self.start_service_time_line.insert(delivery_insert_idx, 0)
		
		# Update service times
		if not self._update_service_times_after_insertion(pick_insert_idx):
			return False, 0.0, 0.0
			
		# Update time cost
		time_cost_diff = self.whole_time_cost - prev_whole_time_cost
		
		# Insert placeholder loads
		self.load_line.insert(pick_insert_idx, 0.0)
		self.load_line.insert(delivery_insert_idx, 0.0)
		
		# Update loads
		if not self._update_loads_after_insertion(pick_insert_idx, delivery_insert_idx):
			return False, 0.0, 0.0
			
		# Insert placeholder distances
		prev_distance = self.distances[-1] if self.distances else 0.0
		self.distances.insert(pick_insert_idx, 0.0)
		self.distances.insert(delivery_insert_idx, 0.0)
		
		# Update distances
		self._update_distances_after_insertion(pick_insert_idx)
		
		current_distance = self.distances[-1] if self.distances else 0.0
		distance_diff = current_distance - prev_distance
		
		return True, distance_diff, time_cost_diff
		
	def try_to_insert_request_optimal(self, request_id: int) -> Tuple[bool, float, float, Optional['Path']]:
		"""Find optimal insertion position for a request with early termination."""
		if self.vehicle_id not in self.meta_obj.requests[request_id].vehicle_set:
			return False, 0.0, 0.0, None
		
		route_len = len(self.route)
		best_cost = float('inf')
		best_path = None
		best_distance = 0.0
		best_time = 0.0
		alpha = self.meta_obj.parameters.alpha
		beta = self.meta_obj.parameters.beta
		for i in range(1, route_len):
			for j in range(i + 1, route_len + 1):
				new_path = self.copy()
				ok, distance_diff, time_cost_diff = new_path.try_to_insert_request(request_id, i, j)
				if ok:
					current_cost = alpha * distance_diff + beta * time_cost_diff
					if current_cost < best_cost:
						best_cost = current_cost
						best_path = new_path
						best_distance = distance_diff
						best_time = time_cost_diff
						if best_cost == 0:  # Early exit if perfect match found
							return True, best_distance, best_time, best_path

		return (True, best_distance, best_time, best_path) if best_path else (False, 0.0, 0.0, None)
	
	def _update_service_times_after_removal(self, start_idx: int) -> bool:
		"""Update service times after removing nodes"""
		for i in range(start_idx, len(self.start_service_time_line)):
			prev_node_id = self.route[i - 1]
			current_node_id = self.route[i]
			
			
			
			new_start_time = max(self.start_service_time_line[i - 1] +
			                     self.meta_obj.nodes[prev_node_id].service_time +
			                     self.meta_obj.vehicle_run_between_nodes_time[self.vehicle_id][prev_node_id][current_node_id],
			                     self.meta_obj.nodes[current_node_id].earliest_service_time)
			
			if new_start_time > self.meta_obj.nodes[current_node_id].latest_service_time:
				return False
			self.start_service_time_line[i] = new_start_time
		return True

	def _update_loads_after_removal(self, pick_idx: int, delivery_idx: int) -> bool:
		"""Update loads after removing pickup and delivery nodes"""
		# Only update loads between pickup and delivery if there are nodes in between
		if pick_idx < delivery_idx - 1:
			for i in range(pick_idx, delivery_idx - 1):
				current_node_id = self.route[i]
				new_load = self.load_line[i - 1] + self.meta_obj.nodes[current_node_id].load
				if new_load > self.meta_obj.vehicles[self.vehicle_id].capacity:
					return False
				self.load_line[i] = new_load
		return True

	def _update_distances_after_removal(self, start_idx: int):
		"""Update distances after removing nodes"""
		for i in range(start_idx, len(self.distances)):
			prev_node_id = self.route[i - 1]
			current_node_id = self.route[i]
			current_distance = self.distances[i - 1] + \
			                   self.meta_obj.distances[prev_node_id][current_node_id]
			self.distances[i] = current_distance

	def try_to_remove_request(self, request_id: int) -> Tuple[float, float]:
		"""Try to remove a request from the path"""
		if self.vehicle_id not in self.meta_obj.requests[request_id].vehicle_set:
			raise PathError(f"Vehicle {self.vehicle_id} not in request {request_id} vehicle set")
			
		pick_node_id = self.meta_obj.requests[request_id].pick_node_id
		delivery_node_id = self.meta_obj.requests[request_id].delivery_node_id
		
		pick_node_idx = self.route.index(pick_node_id)
		delivery_node_idx = self.route.index(delivery_node_id)
		
		if pick_node_idx <= 0 or delivery_node_idx <= 0:
			raise PathError(f"Invalid node indices: pick={pick_node_idx}, delivery={delivery_node_idx}")

		# Remove nodes
		self.route.pop(pick_node_idx)
		self.route.pop(delivery_node_idx - 1)  # Adjust index after first removal

		# Update service times
		prev_whole_time_cost = self.whole_time_cost
		self.start_service_time_line.pop(pick_node_idx)
		self.start_service_time_line.pop(delivery_node_idx - 1)
		
		if not self._update_service_times_after_removal(pick_node_idx):
			raise PathError("Time window violation after removal")
		
		time_cost_diff = prev_whole_time_cost - self.whole_time_cost

		# Update loads
		self.load_line.pop(pick_node_idx)
		self.load_line.pop(delivery_node_idx - 1)
		
		if not self._update_loads_after_removal(pick_node_idx, delivery_node_idx):
			raise PathError("Capacity violation after removal")
		
		# Update distances
		prev_distance = self.distances[-1] if self.distances else 0.0
		self.distances.pop(pick_node_idx)
		self.distances.pop(delivery_node_idx - 1)
		
		self._update_distances_after_removal(pick_node_idx)
		
		current_distance = self.distances[-1] if self.distances else 0.0
		distance_diff = prev_distance - current_distance

		return distance_diff, time_cost_diff

	def get_node_start_service_time(self, node_id: int) -> float:
		"""Get the start service time for a specific node"""
		if node_id not in self.route:
			raise PathError(f"Node {node_id} not found in route")
		node_idx = self.route.index(node_id)
		return self.start_service_time_line[node_idx]
		

File: java-version/src/main/java/com/tc/pdwtw/model/Path.java
package com.tc.pdwtw.model;

import java.util.*;

/**
 * MIT License
 * 
 * Copyright (c) 2024 cauchy1988
 * 
 * Path class representing a vehicle's route through pickup and delivery nodes
 */
public class Path {
  public void setMetaObj(Meta metaObj) {
    this.metaObj = metaObj;
  }

  private Meta metaObj;
    private int vehicleId;
    private List<Integer> route;
    private List<Double> startServiceTimeLine;
    private List<Double> loadLine;
    private List<Double> distances;

    public static class PathError extends RuntimeException {
        public PathError(String message) {
            super(message);
        }
    }

    public static class InsertionResult {
        public final boolean success;
        public final double distanceDiff;
        public final double timeCostDiff;
        public final Path newPath;

        public InsertionResult(boolean success, double distanceDiff, double timeCostDiff, Path newPath) {
            this.success = success;
            this.distanceDiff = distanceDiff;
            this.timeCostDiff = timeCostDiff;
            this.newPath = newPath;
        }
    }

    public Path(int vehicleId, Meta metaObj, boolean needInit) {
        this.metaObj = metaObj;
        this.vehicleId = vehicleId;

        if (needInit) {
            if (!metaObj.getVehicles().containsKey(vehicleId)) {
                throw new PathError("Vehicle " + vehicleId + " not found in meta_obj");
            }

            Vehicle vehicle = metaObj.getVehicles().get(vehicleId);
            int startNodeId = vehicle.getStartNodeId();
            int endNodeId = vehicle.getEndNodeId();

            if (!metaObj.getNodes().containsKey(startNodeId) || !metaObj.getNodes().containsKey(endNodeId)) {
                throw new PathError("Vehicle " + vehicleId + " nodes not found: start=" + startNodeId + ", end=" + endNodeId);
            }

            this.route = new ArrayList<>(Arrays.asList(startNodeId, endNodeId));

            Node startNode = metaObj.getNodes().get(startNodeId);
            Node endNode = metaObj.getNodes().get(endNodeId);
            double earliestTime = startNode.getEarliestServiceTime();
            double arrivalTime = earliestTime + startNode.getServiceTime() + 
                metaObj.getVehicleRunBetweenNodesTime().get(vehicleId).get(startNodeId).get(endNodeId);
            double latestTime = Math.max(arrivalTime, endNode.getEarliestServiceTime());

            if (latestTime > endNode.getLatestServiceTime()) {
                throw new PathError("Time window violation: " + latestTime + " > " + endNode.getLatestServiceTime());
            }

            this.startServiceTimeLine = new ArrayList<>(Arrays.asList(earliestTime, latestTime));
            this.loadLine = new ArrayList<>(Arrays.asList(startNode.getLoad(), startNode.getLoad() + endNode.getLoad()));

            double endDistance = metaObj.getDistances().get(startNodeId).get(endNodeId);
            this.distances = new ArrayList<>(Arrays.asList(0.0, endDistance));
        }
    }

    public Path(int vehicleId, Meta metaObj) {
        this(vehicleId, metaObj, true);
    }

    public Path copy() {
        Path newPath = new Path(this.vehicleId, this.metaObj, false);
        newPath.metaObj = this.metaObj;
        newPath.vehicleId = this.vehicleId;
        newPath.route = new ArrayList<>(this.route);
        newPath.startServiceTimeLine = new ArrayList<>(this.startServiceTimeLine);
        newPath.loadLine = new ArrayList<>(this.loadLine);
        newPath.distances = new ArrayList<>(this.distances);
        return newPath;
    }

    public boolean isPathFree() {
        return route.size() <= 2;
    }

    public double getWholeTimeCost() {
        return startServiceTimeLine.get(startServiceTimeLine.size() - 1) - startServiceTimeLine.get(0);
    }

    public double getWholeDistanceCost() {
        return distances.isEmpty() ? 0.0 : distances.get(distances.size() - 1);
    }

    private boolean updateServiceTimesAfterInsertion(int startIdx) {
        for (int i = startIdx; i < startServiceTimeLine.size(); i++) {
            int prevNodeId = route.get(i - 1);
            int currentNodeId = route.get(i);
            
            double newStartTime = Math.max(
                startServiceTimeLine.get(i - 1) + 
                metaObj.getNodes().get(prevNodeId).getServiceTime() + 
                metaObj.getVehicleRunBetweenNodesTime().get(vehicleId).get(prevNodeId).get(currentNodeId),
                metaObj.getNodes().get(currentNodeId).getEarliestServiceTime()
            );

            if (newStartTime > metaObj.getNodes().get(currentNodeId).getLatestServiceTime()) {
                return false;
            }
            startServiceTimeLine.set(i, newStartTime);
        }
        return true;
    }

    private boolean updateLoadsAfterInsertion(int pickIdx, int deliveryIdx) {
        for (int i = pickIdx; i <= deliveryIdx; i++) {
            int currentNodeId = route.get(i);
            double newLoad = loadLine.get(i - 1) + metaObj.getNodes().get(currentNodeId).getLoad();
            if (newLoad > metaObj.getVehicles().get(vehicleId).getCapacity()) {
                return false;
            }
            loadLine.set(i, newLoad);
        }
        return true;
    }

    private void updateDistancesAfterInsertion(int startIdx) {
        for (int i = startIdx; i < distances.size(); i++) {
            int prevNodeId = route.get(i - 1);
            int currentNodeId = route.get(i);
            double currentDistance = distances.get(i - 1) + 
                metaObj.getDistances().get(prevNodeId).get(currentNodeId);
            distances.set(i, currentDistance);
        }
    }

    public InsertionResult tryToInsertRequest(int requestId, int pickInsertIdx, int deliveryInsertIdx) {
        if (pickInsertIdx >= deliveryInsertIdx) {
            return new InsertionResult(false, 0.0, 0.0, null);
        }

        Request request = metaObj.getRequests().get(requestId);
        int pickNodeId = request.getPickNodeId();
        int deliveryNodeId = request.getDeliveryNodeId();

        route.add(pickInsertIdx, pickNodeId);
        route.add(deliveryInsertIdx, deliveryNodeId);

        double prevWholeTimeCost = getWholeTimeCost();
        startServiceTimeLine.add(pickInsertIdx, 0.0);
        startServiceTimeLine.add(deliveryInsertIdx, 0.0);

        if (!updateServiceTimesAfterInsertion(pickInsertIdx)) {
            return new InsertionResult(false, 0.0, 0.0, null);
        }

        double timeCostDiff = getWholeTimeCost() - prevWholeTimeCost;

        loadLine.add(pickInsertIdx, 0.0);
        loadLine.add(deliveryInsertIdx, 0.0);

        if (!updateLoadsAfterInsertion(pickInsertIdx, deliveryInsertIdx)) {
            return new InsertionResult(false, 0.0, 0.0, null);
        }

        double prevDistance = distances.isEmpty() ? 0.0 : distances.get(distances.size() - 1);
        distances.add(pickInsertIdx, 0.0);
        distances.add(deliveryInsertIdx, 0.0);

        updateDistancesAfterInsertion(pickInsertIdx);

        double currentDistance = distances.isEmpty() ? 0.0 : distances.get(distances.size() - 1);
        double distanceDiff = currentDistance - prevDistance;

        return new InsertionResult(true, distanceDiff, timeCostDiff, this);
    }

    public InsertionResult tryToInsertRequestOptimal(int requestId) {
        Request request = metaObj.getRequests().get(requestId);
        if (!request.getVehicleSet().contains(vehicleId)) {
            return new InsertionResult(false, 0.0, 0.0, null);
        }

        int routeLen = route.size();
        InsertionResult bestResult = null;
        double minCost = Double.MAX_VALUE;
        double alpha = metaObj.getParameters().getAlpha();
        double beta = metaObj.getParameters().getBeta();
        for (int i = 1; i < routeLen; i++) {
            for (int j = i + 1; j <= routeLen; j++) {
                Path newPath = this.copy();
                InsertionResult result = newPath.tryToInsertRequest(requestId, i, j);
                if (result.success) {
                    double currentCost = alpha * result.distanceDiff + beta * result.timeCostDiff;
                    if (currentCost < minCost) {
                        minCost = currentCost;
                        bestResult = result;
                    }
                }
            }
        }

        return bestResult != null
            ? new InsertionResult(true, bestResult.distanceDiff, bestResult.timeCostDiff, bestResult.newPath)
            : new InsertionResult(false, 0.0, 0.0, null);
    }

    public double getNodeStartServiceTime(int nodeId) {
        int nodeIdx = route.indexOf(nodeId);
        if (nodeIdx == -1) {
            throw new PathError("Node " + nodeId + " not found in route");
        }
        return startServiceTimeLine.get(nodeIdx);
    }

    public static class RemovalResult {
        public final double distanceDiff;
        public final double timeCostDiff;

        public RemovalResult(double distanceDiff, double timeCostDiff) {
            this.distanceDiff = distanceDiff;
            this.timeCostDiff = timeCostDiff;
        }
    }

    public RemovalResult tryToRemoveRequest(int requestId) {
        Request request = metaObj.getRequests().get(requestId);
        if (!request.getVehicleSet().contains(vehicleId)) {
            throw new PathError("Vehicle " + vehicleId + " not in request " + requestId + " vehicle set");
        }

        int pickNodeId = request.getPickNodeId();
        int deliveryNodeId = request.getDeliveryNodeId();

        int pickNodeIdx = route.indexOf(pickNodeId);
        int deliveryNodeIdx = route.indexOf(deliveryNodeId);

        if (pickNodeIdx <= 0 || deliveryNodeIdx <= 0) {
            throw new PathError("Invalid node indices: pick=" + pickNodeIdx + ", delivery=" + deliveryNodeIdx);
        }

        // Remove nodes
        route.remove(pickNodeIdx);
        route.remove(deliveryNodeIdx - 1); // Adjust index after first removal

        // Update service times
        double prevWholeTimeCost = getWholeTimeCost();
        startServiceTimeLine.remove(pickNodeIdx);
        startServiceTimeLine.remove(deliveryNodeIdx - 1);

        if (!updateServiceTimesAfterRemoval(pickNodeIdx)) {
            throw new PathError("Time window violation after removal");
        }

        double timeCostDiff = prevWholeTimeCost - getWholeTimeCost();

        // Update loads
        loadLine.remove(pickNodeIdx);
        loadLine.remove(deliveryNodeIdx - 1);

        if (!updateLoadsAfterRemoval(pickNodeIdx, deliveryNodeIdx)) {
            throw new PathError("Capacity violation after removal");
        }

        // Update distances
        double prevDistance = distances.isEmpty() ? 0.0 : distances.get(distances.size() - 1);
        distances.remove(pickNodeIdx);
        distances.remove(deliveryNodeIdx - 1);

        updateDistancesAfterRemoval(pickNodeIdx);

        double currentDistance = distances.isEmpty() ? 0.0 : distances.get(distances.size() - 1);
        double distanceDiff = prevDistance - currentDistance;

        return new RemovalResult(distanceDiff, timeCostDiff);
    }

    private boolean updateServiceTimesAfterRemoval(int startIdx) {
        for (int i = startIdx; i < startServiceTimeLine.size(); i++) {
            int prevNodeId = route.get(i - 1);
            int currentNodeId = route.get(i);

            double newStartTime = Math.max(
                startServiceTimeLine.get(i - 1) + 
                metaObj.getNodes().get(prevNodeId).getServiceTime() + 
                metaObj.getVehicleRunBetweenNodesTime().get(vehicleId).get(prevNodeId).get(currentNodeId),
                metaObj.getNodes().get(currentNodeId).getEarliestServiceTime()
            );

            if (newStartTime > metaObj.getNodes().get(currentNodeId).getLatestServiceTime()) {
                return false;
            }
            startServiceTimeLine.set(i, newStartTime);
        }
        return true;
    }

    private boolean updateLoadsAfterRemoval(int pickIdx, int deliveryIdx) {
        // Only update loads between pickup and delivery if there are nodes in between
        if (pickIdx < deliveryIdx - 1) {
            for (int i = pickIdx; i < deliveryIdx - 1; i++) {
                int currentNodeId = route.get(i);
                double newLoad = loadLine.get(i - 1) + metaObj.getNodes().get(currentNodeId).getLoad();
                if (newLoad > metaObj.getVehicles().get(vehicleId).getCapacity()) {
                    return false;
                }
                loadLine.set(i, newLoad);
            }
        }
        return true;
    }

    private void updateDistancesAfterRemoval(int startIdx) {
        for (int i = startIdx; i < distances.size(); i++) {
            int prevNodeId = route.get(i - 1);
            int currentNodeId = route.get(i);
            double currentDistance = distances.get(i - 1) + 
                metaObj.getDistances().get(prevNodeId).get(currentNodeId);
            distances.set(i, currentDistance);
        }
    }

    // Getters
    public List<Integer> getRoute() { return new ArrayList<>(route); }
    public List<Double> getStartServiceTimeLine() { return new ArrayList<>(startServiceTimeLine); }
    public List<Double> getLoadLine() { return new ArrayList<>(loadLine); }
    public List<Double> getDistances() { return new ArrayList<>(distances); }
    public int getVehicleId() { return vehicleId; }
    public Meta getMetaObj() { return metaObj; }
}